packages/meditrak-server/src/routes/CreateDashboardVisualisation.js

```diff
@@ -0,0 +1,65 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { ObjectValidator } from '@tupaia/utils';
+
+import { CreateHandler } from './CreateHandler';
+import { assertBESAdminAccess } from '../permissions';
+import { constructNewRecordValidationRules } from './utilities';
+
+/**
+ * Handles POST endpoints:
+ * - /dashboardVisualisations
+ */
+
+export class CreateDashboardVisualisation extends CreateHandler {
+  async assertUserHasAccess() {
+    await this.assertPermissions(assertBESAdminAccess);
+  }
+
+  async createReport(transactingModels, reportObject) {
+    const { code, permission_group: permissionGroupName, config } = reportObject;
+    const permissionGroup = await transactingModels.permissionGroup.findOne({
+      name: permissionGroupName,
+    });
+    const report = {
+      code,
+      config,
+      permission_group_id: permissionGroup.id,
+    };
+    return transactingModels.report.create(report);
+  }
+
+  async createDashboardItem(transactingModels, dashboardItemObject) {
+    return transactingModels.dashboardItem.create(dashboardItemObject);
+  }
+
+  async createRecord() {
+    await this.assertPermissions(assertBESAdminAccess);
+
+    return this.models.wrapInTransaction(async transactingModels => {
+      const { report: reportObject, dashboardItem: dashboardItemObject } = this.newRecordData;
+      await this.createReport(transactingModels, reportObject);
+      await this.createDashboardItem(transactingModels, dashboardItemObject);
```

Comment:
Another option is have separate endpoint for saving report and dashboard item. The reason I choose to have 1 combined endpoint is to have 1 transaction so that if it fails saving 1 of them, it shouldn't save the other.

Reply:
I like your reasoning @chris-bes ! I also quite like that we are closely coupling the 'report' and 'dashboardItem' throughout much of logic relating to 'dashboardVisualisations'. Afterall, a dashboardItem requires a report, so it's often nice to treat them as an atomic unit 👍